#!/usr/local/envs/py36/bin python3

import os
import sys
import pandas as pd
from glob import glob


# Import custom functions
from mods import prepareArguments


# Extract variables from configuration file for use within the rest of the pipeline
config = prepareArguments.parsePaths(config)
input_dict = config["inputs"]
output_dict = config["outputs"]
options_dict = config["options"]
ref_dict = config["refs"]
plink_gender_ancestry_QC_dict = config["plink_gender_ancestry_QC"]
grm_QC_dict = config["grm_QC"]
HRC_check_dict = config["HRC_check"]
final_vcf_dict = config["final_vcf"]


# Import individual rules
include: "includes/plink_gender_ancestry_QC.smk"
include: "includes/grm_QC.smk"
include: "includes/HRC_check.smk"
include: "includes/final_vcf.smk"


## Define the chromosomes to be used downstream (post-gcta)
chromosomes = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]


plinkQC_files = []
grm_files = []
vcf_files = []
if os.path.exists(output_dict["output_dir"] + "/pca_sex_checks/ancestry_update_remove.tsv") and os.path.exists(output_dict["output_dir"] + "/pca_sex_checks/check_sex_update_remove.tsv"):
    ancestry_check = pd.read_csv(output_dict["output_dir"] + "/pca_sex_checks/ancestry_update_remove.tsv", sep = "\t")
    sex_check = pd.read_csv(output_dict["output_dir"] + "/pca_sex_checks/check_sex_update_remove.tsv", sep = "\t")
    if ancestry_check["UPDATE/REMOVE/KEEP"].count() == len(ancestry_check) and sex_check["UPDATE/REMOVE/KEEP"].count() == len(sex_check):
        plinkQC_files.append(output_dict["output_dir"] + "/update_sex_ancestry/update_sex.pgen")
        if os.path.exists(output_dict["output_dir"] + "/update_sex_ancestry/uniq_acestries.tsv"):
            ancestry_file = pd.read_csv(output_dict["output_dir"] + "/update_sex_ancestry/uniq_acestries.tsv", sep = "\t")
            if ancestry_file["maf_threshold"].count() == len(ancestry_file):
                ancestry_file.index = ancestry_file["unique_ancestry"]
                grm_files.append(expand(output_dict["output_dir"] + "/grm/{ancestry}_paired_relatedness.tsv", ancestry = ancestry_file["unique_ancestry"]))
                plinkQC_files.append(expand(output_dict["output_dir"] + "/convert_files/{ancestry}/{ancestry}_strand_check-updated-chr{chr}.pvar.temp", ancestry = ancestry_file["unique_ancestry"], chr = chromosomes))
                vcf_files.append(expand(output_dict["output_dir"] + "/vcf/combined_sorted/QC_filtered_sorted_chr{chr}.vcf.gz", chr = chromosomes))
                vcf_files.append(output_dict["output_dir"] + "/vcf/files4submission/QC_filtered_sorted_chr_updated_final.vcf.gz.csi")
                vcf_files.append(output_dict["output_dir"] + "/vcf/files4submission/individual_sexes.txt")
            else:
                logger.info("The minor allele frequency has not been chosen for all the ancestries in update_sex_ancestry/uniq_acestries.tsv. Please update the file to include the minor allele frequency selections with a number between 0 and 1 or NA if you don't want to filter based on a minor allele frequency. Please see https://github.com/sc-eQTLgen-consortium/WG1-pipeline-QC/wiki/1---SNP-Genotype-Imputation#running-the-pipeline---minor-allele-frequency-selections for more details")
        else:
            logger.info("The UPDATE/REMOVE/KEEP column in the pca_sex_checks/ancestry_update_remove.tsv and/or the pca_sex_checks/check_sex_update_remove.tsv file are not completed. Please fill in these selections for the pipeline to continue. Please see https://github.com/sc-eQTLgen-consortium/WG1-pipeline-QC/wiki/1---SNP-Genotype-Imputation#running-the-pipeline---final-qc-and-vcf-production for more details.")
else:
    plinkQC_files.append(output_dict["output_dir"] + "/pca_sex_checks/ancestry_update_remove.tsv")

rule all:
    input:
        plinkQC_files,
        grm_files,
        vcf_files




